docs: add comprehensive dev containers documentation with examples (#18582)
coder/coder: docs/admin/templates/extending-templates/devcontainers.md

       ```terraform
       depends_on       = [module.nodejs]
       ```

   </details>

## Install the Dev Containers CLI

Use the
[devcontainers-cli](https://registry.coder.com/modules/devcontainers-cli) module
to install `@devcontainers/cli` in your workspace:

```terraform
module "devcontainers-cli" {
  count    = data.coder_workspace.me.start_count
  source   = "dev.registry.coder.com/modules/devcontainers-cli/coder"
  agent_id = coder_agent.main.id
}
```

Alternatively, install the devcontainer CLI manually in your base image:

```shell
RUN npm install -g @devcontainers/cli
```

## Define the Dev Container Resource

If you don't use [`git_clone`](#clone-the-repository), point the resource at the folder that contains `devcontainer.json`:

```terraform
resource "coder_devcontainer" "project" {
  count            = data.coder_workspace.me.start_count
  agent_id         = coder_agent.main.id
  workspace_folder = "/home/coder/project"
}
```

## Clone the Repository

This step is optional, but it ensures that the project is present before the dev container starts.

Note that if you use the `git_clone` module, place it before the `coder_devcontainer` resource

> and update or replace that resource to point at `/home/coder/project/${module.git_clone[0].folder_name}` so that it is only defined once:

Reviewer: Order does not matter in terraform.

```terraform
module "git_clone" {
  count    = data.coder_workspace.me.start_count
  source   = "dev.registry.coder.com/modules/git-clone/coder"
  agent_id = coder_agent.main.id
  url      = "https://github.com/example/project.git"

>   base_dir = "/home/coder/project"

Reviewer: ```suggestion
  base_dir = "/home/coder"
```

}

> resource "coder_devcontainer" "project" {

Reviewer: It's unfortunate `folder_name` can't be used here. 😔 I'm really of two-minds using this for the sub-agent name when present but it is what it is.

We support `devcontainer.json` customization to change the name on a per-project basis.

Author: we don't really have that documented, but I think we should def add it in the next iteration

Author: adding: it would be great as part of or in addition to the `Agent naming` section

  count            = data.coder_workspace.me.start_count
  agent_id         = coder_agent.main.id

>   workspace_folder = "/home/coder/project/${module.git_clone[0].folder_name}"

Reviewer: ```suggestion
  workspace_folder = "/home/coder/${module.git_clone[0].folder_name}"
```
This would've resulted in `project/project`.

  depends_on       = [module.git_clone]
}
```

## Dev Container Features

Enhance your dev container experience with additional features.
For more advanced use cases, consult the [advanced dev containers doc](./advanced-dev-containers.md).

### Custom applications

```jsonc
{
  "customizations": {
    "coder": {

>       "apps": {
>         "flask-app": {
>           "command": "python app.py",
>           "icon": "/icon/flask.svg",
>           "subdomain": true,
>           "healthcheck": {
>             "url": "http://localhost:5000/health",
>             "interval": 10,
>             "threshold": 10
>           }
>         }
>       }

Reviewer: ```suggestion
      "apps": [
        {
          "slug": "flask-app",
          "command": "python app.py",
          "icon": "/icon/flask.svg",
          "subdomain": true,
          "healthcheck": {
            "url": "http://localhost:5000/health",
            "interval": 10,
            "threshold": 10
          }
        }
      ]
```
This is the current format, but I think this structure of object-key-as-slug makes more sense. /cc @DanielleMaywood (I guess it's too late to change this?)

PS. Does `healthcheck` and `command` actually make sense together? AFAIK The command is executed on-access meaning it will be unhealthy most of the time. But I may be wrong, have not utilized this feature.

Reviewer 2: Oh this would've made so much more sense, guess it is probably a bit late for that now 🥲

    }
  }
}
```

### Agent naming

Coder names dev container agents in this order:

> 1. `customizations.coder.agent.name` in `devcontainer.json`

Reviewer: ```suggestion
1. `customizations.coder.name` in `devcontainer.json`
```

> 1. `name` in `devcontainer.json`
> 1. Directory name that contains the config
> 1. `devcontainer` (default)

Reviewer: ```suggestion
1. Resource name used in terraform (`resource "coder_devcontainer" "name"`)
1. Project directory name (name of folder containing `devcontainer.json` or `.devcontainer` folder)
1. If project directory name is already taken, the name is expanded to include the parent folder (`/home/coder/some/project` -> `project` (taken) -> `some-project`)
```

  workspace_folder = "/home/coder/frontend"
}

resource "coder_devcontainer" "backend" {
  count            = data.coder_workspace.me.start_count
  agent_id         = coder_agent.main.id
  workspace_folder = "/home/coder/backend"
}
```

## Complete Template Examples

You can test the Coder dev container integration and features with these starter templates.

<details><summary>Docker-based template (privileged)</summary>

This version uses a Docker-in-Docker image, so it works even if the host doesn’t expose a Docker socket.

```terraform
terraform {
  required_providers {
    coder = {
      source = "coder/coder"
    }
    docker = {
      source = "kreuzwerker/docker"
    }
  }
}

provider "coder" {}

data "coder_workspace"       "me"    {}
data "coder_workspace_owner" "me"    {}

resource "docker_container" "workspace" {
  count      = data.coder_workspace.me.start_count
  image      = "codercom/enterprise-base:ubuntu"           # includes Coder agent
  name       = "coder-${data.coder_workspace_owner.me.name}-${lower(data.coder_workspace.me.name)}"

  # share host Docker
  volumes {
    host_path      = "/var/run/docker.sock"
    container_path = "/var/run/docker.sock"

>   }

Reviewer: This is a very bad idea. All workspaces will try to take over each others' devcontainers.

  agent_id = coder_agent.main.id
}

module "git_clone" {
  count    = data.coder_workspace.me.start_count
  source   = "dev.registry.coder.com/modules/git-clone/coder"
  agent_id = coder_agent.main.id
  url      = "https://github.com/coder/coder.git"
  base_dir     = "/home/coder/project"
}

resource "coder_devcontainer" "project" {
  count            = data.coder_workspace.me.start_count
  agent_id         = coder_agent.main.id
  workspace_folder = "/home/coder/project/${module.git_clone[0].folder_name}"
  depends_on       = [module.git_clone]
}

resource "kubernetes_pod" "workspace" {
  count = data.coder_workspace.me.start_count

  metadata {
    name       = "coder-${data.coder_workspace_owner.me.name}-${lower(data.coder_workspace.me.name)}"
    namespace = "coder-workspaces"
  }

  spec {
    container {
      name  = "main"
      image = "codercom/enterprise-base:ubuntu"

      security_context { privileged = true }  # or use Sysbox / rootless
      env { name = "CODER_AGENT_TOKEN" value = coder_agent.main.token }
    }
  }
}
```

</details>

## Troubleshoot Common Issues

### Disable dev containers integration

> To disable the dev containers integration in your workspace, add the `CODER_AGENT_DEVCONTAINERS_ENABLE` environment variable to your existing `coder_agent` block:

Reviewer: This is incorrect, the env must be set _before_ the agent is started, i.e. on the Docker container. We could potentially implement this behavior but it's currently not supported (/cc @DanielleMaywood thoughts?)

Author: changed to

~~~md
To disable the dev containers integration in your workspace, set the `CODER_AGENT_DEVCONTAINERS_ENABLE= "false"` environment variable.
~~~
